subsys/bluetooth/audio/shell/bap.c

```diff
@@ -2330,6 +2349,150 @@ static struct bt_le_scan_cb bap_scan_cb = {
+	struct net_buf *buf;
+	struct shell_stream *sh_stream;
+};
+
+K_MEM_SLAB_DEFINE(lc3_data_slab, sizeof(struct lc3_data), CONFIG_BT_ISO_RX_BUF_COUNT,
+		  __alignof__(struct lc3_data));
+
+static int16_t lc3_rx_buf[LC3_MAX_NUM_SAMPLES];
+static K_FIFO_DEFINE(lc3_in_fifo);
+
+static int init_lc3_decoder(struct shell_stream *sh_stream)
+{
+	if (sh_stream == NULL) {
+		shell_error(ctx_shell, "invalid stream to init LC3");
+		return -EINVAL;
+	}
+
+	if (sh_stream->lc3_decoder != NULL) {
+		shell_error(ctx_shell, "Already initialized");
+		return -EALREADY;
+	}
+
+	if (sh_stream->lc3_freq_hz == 0 || sh_stream->lc3_frame_duration_us == 0) {
+		shell_error(ctx_shell, "Invalid freq (%u) or frame duration (%u)",
+			    sh_stream->lc3_freq_hz, sh_stream->lc3_frame_duration_us);
+
+		return -EINVAL;
+	}
+
+	shell_print(ctx_shell,
+		    "Initializing the LC3 decoder with %u us duration and %u Hz frequency",
+		    sh_stream->lc3_frame_duration_us, sh_stream->lc3_freq_hz);
+	/* Create the decoder instance. This shall complete before stream_started() is called. */
+	sh_stream->lc3_decoder = lc3_setup_decoder(sh_stream->lc3_frame_duration_us,
+						   sh_stream->lc3_freq_hz, 0, /* No resampling */
+						   &sh_stream->lc3_decoder_mem);
+	if (sh_stream->lc3_decoder == NULL) {
+		shell_error(ctx_shell, "Failed to setup LC3 decoder - wrong parameters?\n");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static void lc3_decoder_stream_clear(struct shell_stream *sh_stream)
+{
+	sh_stream->lc3_decoder = NULL;
+}
+
+static bool decode_frame(struct net_buf *buf, struct shell_stream *sh_stream, size_t frame_cnt)
+{
+	const size_t total_frames = sh_stream->lc3_chan_cnt * sh_stream->lc3_frame_blocks_per_sdu;
+	const uint16_t octets_per_frame = sh_stream->lc3_octets_per_frame;
+	void *data;
+	int err;
+
+	if (buf->len == 0U) {
+		data = NULL; /* perform PLC */
+
+		if ((sh_stream->decoded_cnt % recv_stats_interval) == 0) {
+			shell_print(ctx_shell, "[%zu]: Performing PLC", sh_stream->decoded_cnt);
+		}
+	} else {
+		data = net_buf_pull_mem(buf, octets_per_frame);
+
+		if ((sh_stream->decoded_cnt % recv_stats_interval) == 0) {
+			shell_print(ctx_shell, "[%zu]: Decoding frame of size %u (%u/%u)",
+				    sh_stream->decoded_cnt, octets_per_frame, frame_cnt + 1,
+				    total_frames);
+		}
+	}
+
+	err = lc3_decode(sh_stream->lc3_decoder, data, octets_per_frame, LC3_PCM_FORMAT_S16,
+			 lc3_rx_buf, 1);
+	if (err < 0) {
+		shell_error(ctx_shell, "Failed to decode LC3 data (%u/%u - %u/%u)", frame_cnt + 1,
+			    total_frames, octets_per_frame * frame_cnt, buf->len);
+		return false;
+	}
+
+	return true;
+}
+
+static size_t decode_frame_block(struct net_buf *buf, struct shell_stream *sh_stream,
+				 size_t frame_cnt)
+{
+	const uint8_t chan_cnt = sh_stream->lc3_chan_cnt;
+
+	for (uint8_t j = 0U; j < chan_cnt; j++) {
+		if (!decode_frame(buf, sh_stream, frame_cnt)) {
+			break;
+		}
+
+		frame_cnt++;
+	}
+
+	return frame_cnt;
```

Comment:
This function should return how many channnel decode success in one frame block, rather than the total frame_cnt.

Reply:
Ah, yes, good catch